### PR TITLE
Refactor optimize-1q-gates

### LINKED ISSUE #2762: Single qubit gate optimization ignores basis gates
### Information

- **Qiskit Terra version**: master
- **Python version**: 3.7
- **Operating system**: MacOS

### What is the current behavior?

If I have a circuit of single-qubit gates that I want to combine into u3 gates, using the transpiler with `optimization_level=1` expands `u3` gates back into `u1, u2, u3` gates.

### Steps to reproduce the problem

```python
from qiskit import QuantumCircuit
from qiskit.compiler import transpile

# Create single-qubit gate circuit
circ = QuantumCircuit(2)
circ.x(0)
circ.s(0)
circ.h(1)
circ.barrier([0, 1])
circ.z(0)
circ.t(0)
circ.h(1)
circ.sdg(1)
circ.barrier([0, 1])
print(circ)

        ┌───┐┌───┐ ░ ┌───┐ ┌───┐  ░ 
q_0: |0>┤ X ├┤ S ├─░─┤ Z ├─┤ T ├──░─
        ├───┤└───┘ ░ ├───┤┌┴───┴┐ ░ 
q_1: |0>┤ H ├──────░─┤ H ├┤ Sdg ├─░─
        └───┘      ░ └───┘└─────┘ ░ 

# Try and fuse gates
new_circ = transpile(circ, basis_gates=['u3'], optimization_level=1)
new_circ.draw()

        ┌──────────────────────────┐ ░         ┌───────────┐         ░ 
q_0: |0>┤ U3(3.1416,1.5708,3.1416) ├─░─────────┤ U1(3.927) ├─────────░─
        └─────┬──────────────┬─────┘ ░ ┌───────┴───────────┴───────┐ ░ 
q_1: |0>──────┤ U2(0,3.1416) ├───────░─┤ U3(1.5708,-1.5708,3.1416) ├─░─
              └──────────────┘       ░ └───────────────────────────┘ ░ 
```

### What is the expected behavior?

The output circuit should only contain `u3` single qubit gates

### Suggested solutions

I think after unrolling to the `u3` basis the `Optimize1qGates` transpiler pass converts u3 gates back into u1 or u2 if they have parameters corresponding to those types of gates

One possible solution would be to add some options to the `Optimize1qGates` pass to not up-convert gates (ie allow u1,u2->u3 but not u3->u1,u2). Another useful option here would be to allow an option to leave gates equivalent to identities in the circuit as `u1(0)` or `u3(0,0,0)`

### LINKED ISSUE #3360: Transpiler won't remove Gate->Inverse pairs when basis_gates is None
Reported by @ewoutvandenberg , unless a basis is specified, the transpiler does not remove pairs of gate-inverse gate. e.g.

```
>>> qc = qk.QuantumCircuit(1)
>>> qc.s(0)
>>> qc.sdg(0)
>>> print(qc)
        ┌───┐┌─────┐
q_0: |0>┤ S ├┤ Sdg ├
        └───┘└─────┘
>>> print(qk.transpile(qc))
        ┌───┐┌─────┐
q_0: |0>┤ S ├┤ Sdg ├
        └───┘└─────┘
>>> print(qk.transpile(qc, basis_gates=['u3']))

q_0: |0>
```

### LINKED ISSUE #3433: Incorrect transpilation of phase-gadgets with small angles on optimisation level 2 and 3
### Information

- **Qiskit Terra version**: 0.10.0
- **Python version**: 3.6.9
- **Operating system**: MacOS Mojave

### What is the current behavior?

If there is a ZZ pauli exponential with a small angle, Qiskit will remove the CXs and leave the small angle. (It will also do this if the Xs are not controlled)

### Steps to reproduce the problem

Create a `QuantumCircuit` with an Rz nested between the targets of two CX gates. See this qasm:

```
OPENQASM 2.0;
include "qelib1.inc";

qreg q[2];
cx q[1],q[0];
rz(3.978863630138309e-07*pi) q[0];
cx q[1],q[0];
```

Then load the qasm into qiskit and run an optimisation pass:
```
from qiskit import QuantumCircuit
from qiskit.compiler import transpile
from qiskit.transpiler import CouplingMap

qsc = QuantumCircuit.from_qasm_file("test_qiskit.qasm")
basis_gates = ['u1','u2','u3','cx']
cm = None

qsc2 = transpile(qsc,basis_gates=basis_gates,coupling_map=cm,optimization_level=2)
print(qsc2)
```

### What is the expected behavior?

Either the circuit should be left unchanged or, if the angle is considered to be too small to be worth leaving in and it should be approximated away, then remove the entire phase-gadget. I'm not sure whether this was a design choice or a bug, as I don't really understand how it is possible to write this behaviour by accident.

### Suggested solutions

See above.

### LINKED ISSUE #3563: optimize_1q_gates should handle chain of arbitrary controlled gates
From https://github.com/Qiskit/qiskit-terra/issues/3453
> the optimize_1q_gates should be able to handle optimizations where the chain of single-qubit gates are controlled by the same control lines

### PR DESCRIPTION
**Pull request by ajavadia**
Fixes #3360 
Fixes #2762 
Fixes #3433 

This commit deprecates the `Optimize1qGates` pass in favor of a new pass `Collapse1qChains`. This pass merges and collapses uninterrupted chains of single qubit gates into a 2x2 matrix operator. A follow-on invocation of the `UnitarySynthesis` can then rewrite this over any basis. This means that single qubit gates can now be optimized over any basis supported by our `OneQubitEulerDecomposer` class:
`U`, `ZYZ`, `ZXZ`, `XYX`, `PSX`, `RR`, `U3`, `U1U2U3` 
(the last 2 are kept to preserve the ability to optimize circuits for backends that have the older IQX gateset).

This pass also works on chains of length 1. So for example a single U can be written in terms of a simpler Phase+SX if possible.

Later follow-ups needed:
- [ ] make this work on controlled gates too (#3563)
- [ ] merge Collect2qBlocks and ConsolidateBlocks into a single pass similar to this one, since they are always invoked together
- [ ] make these chain collapsing passes work on the DAGDependency as it creates more discovery opportunities

### CODE REVIEW

File: qiskit/transpiler/passes/simplify_u3.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                op = node.op.base_gate
+
+            if (isinstance(op, U3Gate)):
+                theta, phi, lam = op.params
+
+                theta = np.mod(theta, 2*np.pi)
+                phi = np.mod(phi, 2*np.pi)
+                lam = np.mod(lam, 2*np.pi)
+
+                new_op = U3Gate(theta, phi, lam)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/simplify_u3.py]: Maybe move this line to the `else` case for the following `elif`?

File: qiskit/transpiler/passes/simplify_u3.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                elif np.isclose(theta, np.pi/2, atol=DEFAULT_ATOL):
+                    new_op = U2Gate(phi, lam)
+
+                elif np.isclose(theta, 3*np.pi/2, atol=DEFAULT_ATOL):
+                    phi = np.mod(phi+np.pi, 2*np.pi)
+                    lam = np.mod(lam+np.pi, 2*np.pi)
+                    new_op = U2Gate(phi, lam)
+
+            if new_op is None:
+                dag.remove_op_node(node)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/simplify_u3.py]: I'm executing in my head, so I might be wrong. if `op` is not a `U3Gate` (therefore `new_op` is `None`, then it is removed? Is that the intention?

Comment (ajavadia) [reply qiskit/transpiler/passes/simplify_u3.py]: yeah this is not done yet

File: qiskit/circuit/add_control.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 from qiskit.extensions import UnitaryGate
 
 
-def add_control(operation, num_ctrl_qubits, label):
+def add_control(operation, num_ctrl_qubits, label=None):
```
Comment (ewinston) [inline qiskit/circuit/add_control.py]: `add_control` was not meant to be used directly but rather through,
`operation.control(num_ctrl_qubits)` where `label` is optional.

Comment (ajavadia) [reply qiskit/circuit/add_control.py:21]: this is updated on master now

**Comment by mtreinish** [comment]
I'm going to close this PR, it's been idle for 2 years and the 1q optimization passes have changed quite a bit since this was originally written. But please feel free to re-open this if I'm missing something

**Pull request closed without merging**